File: 0787-cheapest-flights-within-k-stops/0787-cheapest-flights-within-k-stops.py
# from collections import *

# class Solution:
#     def findCheapestPrice(self, n: int, flights: List[List[int]], src: int, dst: int, k: int) -> int:
#         flightDict = defaultdict(list)
#         for fr, to, pr in flights:
#             flightDict[fr].append((pr, to))
#         hq = [(0, src, -1)]
#         while len(hq) > 0:
#             price, current, count = heappop(hq)
#             if current == dst:
                
#                 return price
            
#             if count + 1 <= k:
#                 for pr, to in flightDict[current]:
#                     heappush(hq, (price + pr, to, count + 1))
        
#         return -1
import collections
import heapq

class Solution:
    def findCheapestPrice(self, n: int, flights: List[List[int]], src: int, dst: int, k: int) -> int:
        flights_dict = collections.defaultdict(list)
        for fr, to , pr in flights:
            flights_dict[fr].append([pr, to])
        prices = collections.defaultdict(lambda :float('inf'))
        nofs = collections.defaultdict(int)
        hq = [[0, src, 0]]
        while len(hq) > 0:
            pr, to, nof = heapq.heappop(hq)
            if to == dst:
                    
                return pr
            if (pr < prices[to] or nof < nofs[to]) and nof < k + 1:
                prices[to] = pr
                nofs[to] = nof
                for npr, nto in flights_dict[to]:
                    heapq.heappush(hq, [pr + npr, nto, nof + 1])
                    
        return -1

File: 0787-cheapest-flights-within-k-stops/0787-cheapest-flights-within-k-stops.java
class Solution {
    public int findCheapestPrice(int n, int[][] flights, int src, int dst, int k) {
        int[] result = new int[n];
        Arrays.fill(result, -1);
        result[src] = 0;

        for (int i = 0; i < k + 1; i++) {
            int[] newResult = Arrays.copyOfRange(result, 0, result.length);

            for (int[] flight: flights) {
                int from = flight[0];
                int to = flight[1];
                int time = flight[2];

                if (result[from] < 0) {
                    continue;
                }

                newResult[to] = Arrays.stream(new int[] {result[to], newResult[to], result[from] + time})
                    .filter((a) -> a >= 0)
                    .min()
                    .getAsInt();
            }

            result = newResult;
        }

        return result[dst];
    }
}